File: swift/list-ops/Sources/ListOps/ListOps.swift
import Foundation

class ListOps {

	public static func append<T>(_ list: [T], _ items: [T]) -> [T] {
		return list + items
	}

	public static func concat<T>(_ list: [T]...) -> [T] {
		var res = [T]();

		for item in list {
		  for e in item {
			res.append(e)
		  }
		}
		return res
	}

	static func map<T>(_ list: [T], _ predicate: (T) -> T) -> [T] {
		var res = [T]();
		for item in list {
			res.append(predicate(item))
		}
		return res
	}

	static func filter<T>(_ list: [T], _ predicate: (T) -> Bool) -> [T] {
		var res = [T]();
		for item in list {
			if predicate(item) {
				res.append(item)
			}
		}
		return res
	}
	
	static func foldLeft<T>(_ list: [T], accumulated: T, combine: (T, T) -> T) -> T {
		var res = accumulated
		for item in list {
			res = combine(res, item)
		}
		return res
	}
	
	static func foldRight<T>(_ list: [T], accumulated: T, combine: (T, T) -> T) -> T {
		var res = accumulated
		for item in self.reverse(list) {
			res = combine(item, res)
		}
		return res
	}
	
	static func length<T>(_ list: [T]) -> Int {
		var len = 0
		for _ in list {
			len += 1
		}
		return len
	}
	
	static func reverse<T>(_ list: [T]) -> [T] {
		var res = [T]();
		var i = list.count - 1
		while i >= 0 {
			res.append(list[i])
			i -= 1
		}
		return res
	}
}

File: csharp/list-ops/ListOps.cs
using System;
using System.Collections.Generic;
using System.Linq;

public static class ListOps
{
    public static int Length<T>(List<T> input)
    {
        var res = 0;
        foreach (var _ in input) {
            res++;
        }
        return res;
    }

    public static List<T> Reverse<T>(List<T> input)
    {
        var inputLength = ListOps.Length(input);
        var res = new List<T>();
        for (int i = inputLength - 1; i >= 0; i--) {
            res = ListOps.Append(res, new List<T>(){ input[i] });
        }
        return res;
    }

    public static List<TOut> Map<TIn, TOut>(List<TIn> input, Func<TIn, TOut> map)
    {
        var res = new List<TOut>();
        for (int i = 0; i < ListOps.Length(input); i++) {
            res = ListOps.Append(res, new List<TOut>(){ map(input[i]) });
        }
        return res;
    }

    public static List<T> Filter<T>(List<T> input, Func<T, bool> predicate)
    {
        var inputLength = ListOps.Length(input);
        var res = new List<T>();
        for (int i = 0; i < ListOps.Length(input); i++) {
            if (predicate(input[i])) {
                res = ListOps.Append(res, new List<T>(){ input[i] });
            }
        }
        return res;
    }

    public static TOut Foldl<TIn, TOut>(List<TIn> input, TOut start, Func<TOut, TIn, TOut> func)
    {
        TOut res = start;
        for (int i = 0; i < ListOps.Length(input); i++) {
            res = func(res, input[i]);
        }
        return res;
    }

    public static TOut Foldr<TIn, TOut>(List<TIn> input, TOut start, Func<TIn, TOut, TOut> func)
    {
        TOut res = start;
        for (int i = ListOps.Length(input) - 1; i >= 0; i--) {
            res = func(input[i], res);
        }
        return res;
    }

    public static List<T> Concat<T>(List<List<T>> input)
    {
        var res = new List<T>();
        foreach (var list in input) {
            res = ListOps.Append(res, list);
        }
        return res;
    }

    public static List<T> Append<T>(List<T> left, List<T> right)
    {
        var leftLength = ListOps.Length(left);
        var rightLength = ListOps.Length(right);
        var res = new T[leftLength + rightLength];
        
        for (int i = 0; i < leftLength; i++) {
            res[i] = left[i];
        }
        
        for (int i = leftLength; i < leftLength + rightLength; i++) {
            res[i] = right[i - leftLength];
        }
        return res.ToList();
    }
}